File: packages/engine/src/cpus/6502/instructions/rti.test.js
import { test, expect } from 'vitest'
import { BufferMapper } from '../../../buffer_mapper'
import { CPU } from '..'

test('rti', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.sp = 0xff
  cpu.interruptDisabled = false
  cpu.zero = true
  mmu.data = [0x0] // BRK
  mmu.data[0xfffe] = 0x04
  mmu.data[0xffff] = 0xac
  const g = cpu.run()
  for (let i = 0; i < 7; i++) g.next()
  expect(cpu.pc).toBe(0x4ac)
  expect(mmu.data[0x1ff]).toBe(0)
  expect(mmu.data[0x1fe]).toBe(2)
  expect(mmu.data[0x1fd]).toBe((cpu.status | 0x10) & ~0x4)
  expect(cpu.interruptDisabled).toBeTruthy()

  mmu.data[0x4ac] = 0x40 // RTI
  cpu.zero = false
  for (let i = 0; i < 6; i++) g.next()
  expect(cpu.pc).toBe(2)
  expect(cpu.zero).toBeTruthy()
  expect(cpu.interruptDisabled).toBeFalsy()
})

File: packages/engine/src/cores/famibob/cpu/instructions/rti.test.zig
const testing = @import("../testing/setup.zig");

test "RTI" {
    var cpu = testing.createCPU();
    cpu.sp = 0xff;
    cpu.i = false;
    cpu.z = true;
    cpu.write(0, 0);
    cpu.write(0xfffe, 4);
    cpu.write(0xffff, 0xac);
    for (0..7) |_| {
        cpu.process();
    }
    try testing.expectEqual(0x4ac, cpu.pc);
    try testing.expectEqual(0, cpu.read(0x1ff));
    try testing.expectEqual(2, cpu.read(0x1fe));
    try testing.expectEqual((cpu.getP() | 0x10) & ~@as(u8, 0x4), cpu.read(0x1fd));
    try testing.expectEqual(true, cpu.i);

    cpu.write(0x4ac, 0x40);
    cpu.z = false;
    for (0..6) |_| {
        cpu.process();
    }
    try testing.expectEqual(2, cpu.pc);
    try testing.expectEqual(true, cpu.z);
    try testing.expectEqual(false, cpu.i);
}
